packages/terra-application/src/page/PageHeader.jsx

```diff
@@ -0,0 +1,220 @@
+import classNames from 'classnames/bind';
+import IconLeft from 'terra-icon/lib/icon/IconLeft';
+import IconRollup from 'terra-icon/lib/icon/IconRollup';
+import Popup from 'terra-popup';
+
+import ActionMenu, { ActionMenuItem } from '../action-menu';
+import { useTransientPresentationState } from '../utils/transient-presentations';
+import useElementSize, { breakpointFilter } from '../shared/useElementSize';
+import PageContainerContext from '../page-container/PageContainerContext';
+import { ApplicationIntlContext } from '../application-intl';
+
+import PageHeaderButton from './PageHeaderButton';
+
+import styles from './PageHeader.module.scss';
+
+const cx = classNames.bind(styles);
+
+const propTypes = {
+  /**
+   * A PageActions element containing action definitions to render as controls
+   * within the header.
+   */
+  actions: PropTypes.element,
+  /**
+   * A unique identifier for the PageHeader. Applied to internal components
+   * as an id attribute and thus must be globally unique.
+   */
+  id: PropTypes.string.isRequired,
+  /**
+   * The label for the Page that the header is being rendered for.
+   */
+  label: PropTypes.string,
+  /**
+   * A NotificationBanners element type (as provided by the useNotificationBanners
+   * hook) to be rendered alongside the other header contents.
+   */
+  NotificationBanners: PropTypes.elementType,
+  /**
+   * A function to be executed upon selection of the Page's back button. If not
+   * provided, the back button will not be rendered.
+   */
+  onSelectBack: PropTypes.func,
+  /**
+   * A PageToolbar element to render alongside the other header contents.
+   */
+  toolbar: PropTypes.element,
+};
+
+const PageHeader = ({
+  actions, id, label, NotificationBanners, onSelectBack, toolbar,
+}) => {
+  const intl = React.useContext(ApplicationIntlContext);
+  const pageContainerContext = React.useContext(PageContainerContext);
+
+  const headerContainerRef = React.useRef();
+  const moreActionsButtonRef = React.useRef();
+
+  const [showMenu, setShowMenu] = useTransientPresentationState(false);
+  const { activeBreakpoint } = useElementSize(headerContainerRef, breakpointFilter);
+
+  // We filter out any falsy action values that may have been provided
+  // due to inline boolean logic.
+  let validActions = [];
+  if (actions) {
+    validActions = React.Children.toArray(actions.props.children).filter(child => child);
+  }
+
+  if (validActions.length > 3) {
+    throw new Error(`[terra-application] ${label} cannot render more than three actions.`);
+  }
+
+  const actionsAreCollapsed = activeBreakpoint === 'tiny' && validActions.length > 1;
+
+  function renderActionButtons() {
+    return validActions.map((action) => (
+      <PageHeaderButton
+        key={action.key}
+        refCallback={action.props.refCallback}
+        icon={action.props.icon}
+        ariaLabel={action.props.label}
+        onSelect={action.props.onSelect}
+      />
+    ));
+  }
+
+  function renderMenuButton() {
```

Comment:
what are your thoughts on further abstracting these render functions into their own components? Looks like we have `ActionButtons`,  `MenuButton`, `Menu`, `Toolbar`, `StartActions`, `EndActions`. Extracting these out will remove nested rendering and make this component easier to follow

Reply:
Separate components would result in a lot of props being passed around and seems like overkill (for what would amount to a single-use component). But I might be able to make this more straightforward, I'll play with it a bit.